applications/FluidDynamicsApplication/custom_elements/embedded_fluid_element_discontinuous.cpp

```diff
@@ -0,0 +1,975 @@
+    // because rData.ElementSize might still be uninitialized: some data classes define it at the Gauss point.
+    double h = ElementSizeCalculator<Dim,NumNodes>::MinimumElementSize(this->GetGeometry());
+    const double tolerance = std::pow(1e-3 * h, Dim-1);
+    this->NormalizeInterfaceNormals(rData.PositiveInterfaceUnitNormals, tolerance);
+    this->NormalizeInterfaceNormals(rData.NegativeInterfaceUnitNormals, tolerance);
+}
+
+template <class TBaseElement>
+void EmbeddedFluidElementDiscontinuous<TBaseElement>::NormalizeInterfaceNormals(
+    typename EmbeddedDiscontinuousElementData::InterfaceNormalsType& rNormals,
+    double Tolerance) const
+{
+    for (unsigned int i = 0; i < rNormals.size(); ++i) {
+        double norm = norm_2(rNormals[i]);
+        rNormals[i] /= std::max(norm,Tolerance);
+    }
+}
+
+template <class TBaseElement>
+void EmbeddedFluidElementDiscontinuous<TBaseElement>::AddNormalPenaltyContribution(
+    MatrixType& rLHS,
+    VectorType& rRHS,
+    const EmbeddedDiscontinuousElementData& rData) const
+{
+    // Obtain the previous iteration velocity solution
+    array_1d<double,LocalSize> values;
+    this->GetCurrentValuesVector(rData, values);
+
+    // If there is embedded velocity, substract it to the previous iteration solution
+    if (this->Has(EMBEDDED_VELOCITY)){
+        const array_1d<double, 3 >& embedded_vel = this->GetValue(EMBEDDED_VELOCITY);
+        array_1d<double, LocalSize> embedded_vel_exp(LocalSize, 0.0);
+
+        for (unsigned int i = 0; i < NumNodes; ++i){
+            for (unsigned int comp = 0; comp < Dim; ++comp){
+                embedded_vel_exp(i*BlockSize + comp) = embedded_vel(comp);
+            }
+        }
+
+        noalias(values) -= embedded_vel_exp;
+    }
+
+    // Compute the Nitsche normal imposition penalty coefficient
+    const double pen_coef = this->ComputeNormalPenaltyCoefficient(rData);
+
+    // Compute the positive side LHS and RHS contributions
+    const unsigned int number_of_positive_interface_integration_points = rData.PositiveInterfaceWeights.size();
+    for (unsigned int g = 0; g < number_of_positive_interface_integration_points; ++g) {
+        // Get the Gauss pt. data
+        const double weight = rData.PositiveInterfaceWeights[g];
+        const auto aux_N = row(rData.PositiveInterfaceN, g);
+        const auto &aux_unit_normal = rData.PositiveInterfaceUnitNormals[g];
+
+        // Compute the Gauss pt. LHS contribution
+        for (unsigned int i = 0; i < NumNodes; ++i){
+            for (unsigned int j = 0; j < NumNodes; ++j){
+                for (unsigned int m = 0; m < Dim; ++m){
+                    const unsigned int row = i * BlockSize + m;
+                    for (unsigned int n = 0; n < Dim; ++n){
+                        const unsigned int col = j * BlockSize + n;
+                        rLHS(row, col) += pen_coef*weight*aux_N(i)*aux_unit_normal(m)*aux_unit_normal(n)*aux_N(j);
```

Comment:
it may be worth to create a tmp double to avoid calculations 
double tmp = pen_coef*weight*aux_N(i)*aux_unit_normal(m)*aux_unit_normal(n)

Reply:
Done.